### ISSUE TITLE
Replace invalid characters in file names

### ISSUE THREAD

**Issue by hugbug** [nzbget/nzbget#209]
NZBGet already replaces invalid characters in file names when reading file names from nzb-file. During download however the final file names are read from article data. Sometimes when the files were uploaded from a Unix machine it may happen that the file names contain characters not allowed on Windows, such as `:`(colon). Currently NZBGet uses the file names from article data directly without validating.

Also see forum topic [Error during download](http://forum.nzbget.net/viewtopic.php?f=3&t=2503).
### In this issue:
- [x] replace invalid characters in file names which were read from article data;
- [x] make necessary changes in par2-module to ensure par-repair works correctly with changed file names;
- [x] make necessary changes in par-renamer.

**Comment by Safihre** [comment]
No idea if you are aware of it, but there's also a set of reserved words in Windows that file names cannot start with. 
Any of these followed by a dot (ex `con.xxxxxxxx`):

```
_DEVICES = ('con', 'prn', 'aux', 'nul',
            'com1', 'com2', 'com3', 'com4', 'com5', 'com6', 'com7', 'com8', 'com9',
            'lpt1', 'lpt2', 'lpt3', 'lpt4', 'lpt5', 'lpt6', 'lpt7', 'lpt8', 'lpt9')
```

There is 1 poster out there that posts the show "Con.Man" from a Linux system, but on Windows this goes very wrong. There's nothing really to do about it even if you prefix it with something during Adding, Downloading and Par2, since you can't control what comes out of unrar.

Example NZB, causes NZBGet to use 100% CPU on Windows even before finishing:
[_Con.Man.S01E04.Retarding.It.All.Up.720p.WEB-DL.DD5.1.H.264-EsQ (1).nzb.zip](https://github.com/nzbget/nzbget/files/235966/_Con.Man.S01E04.Retarding.It.All.Up.720p.WEB-DL.DD5.1.H.264-EsQ.1.nzb.zip)

**Comment by hugbug** [comment]
Another nice thing about Windows :wink: 
Thank you.
I'll see what I can do about this.

**Comment by hugbug** [comment]
> _DEVICES = ('con', 'prn', 'aux', 'nul',
>             'com1', 'com2', 'com3', 'com4', 'com5', 'com6', 'com7', 'com8', 'com9',
>             'lpt1', 'lpt2', 'lpt3', 'lpt4', 'lpt5', 'lpt6', 'lpt7', 'lpt8', 'lpt9')

Created a separate issue #210.

**Comment by hugbug** [comment]
@Safihre :

> There is 1 poster out there that posts the show "Con.Man" from a Linux system, but on Windows this goes very wrong. There's nothing really to do about it even if you prefix it with something during Adding, Downloading and Par2, since you can't control what comes out of unrar.

Actually it works fine in NZBGet with the latest develop (after #210-fix). The solution consists of:
- make sure directory and all files do not start with reserved words (rename during or after download);
- when executing `unrar` pass the full destination directory in extended format (using `\\?\`-notation);
- validate and adjust file names after unpack for user to have easy access to files from all apps (not only from apps supporting extended path format).

At the end I have directory "_Con.Man.S01E04.Retarding.It.All.Up.720p.WEB-DL.DD5.1.H.264-EsQ" with file "_Con.Man.S01E04.Retarding.It.All.Up.720p.WEB-DL.DD5.1.H.264-EsQ.mkv" which plays in standard player.

**Comment by Safihre** [comment]
Glad you were able to fix it, we (I'm from SABnzbd) just decided not to go through all the trouble and hope this situation doesn't happen too often :sweat_smile: 
Only reported twice over the past many years, but still: Windows........

Just happend to see this issue/commit here when trying to learn more about how other NZB grabbers process the article data from start to finish and what I could learn to push most out of Python.
